### ISSUE TITLE
storage: second time writer.Close() is called it hangs

### ISSUE THREAD

**Issue by mfederw** [googleapis/google-cloud-go#3313]
**Client**

e.g. PubSub

**Environment**

Running on a linux machine

**Go Environment**

GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/dwshome/mike/.cache/go-build"
GOENV="/dwshome/mike/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/dwshome/mike/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go-1.14.2/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go-1.14.2/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/dwshome/mike/src/mcm/copy/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build294052551=/tmp/go-build -gno-record-gcc-switches"
$ go version
go version go1.14.2 linux/amd64$ go env

**Code**
func (gcep *gcEndpoint) putObjectBytes(objName string, buf []byte) error {
        obj := gcep.bucket.Object(objName)
        ctx := context.TODO()
        w := obj.NewWriter(ctx)

        if gcep.key == nil {
                buffer := bytes.NewBuffer(buf)

                _, err := io.Copy(w, buffer)
                if err != nil {
                        // w.Close() and handle error
                        return errors.New("putObjectBytes Copy " + err.Error())
                }
        } else {
                cipherBuf, err := encrypt.Encrypt(buf, *gcep.key)
                if err != nil {
                         // w.Close() and handle error
                        return errors.New("putObjectBytes Encrypt: " + err.Error())
                }
                buffer := bytes.NewBuffer(cipherBuf)
                _, err = io.Copy(w, buffer)
                if err != nil {
                        // w.Close() and handle error
                        return errors.New("putObject Encrypted Copy: " + err.Error())
                }
        }
        fmt.Printf("Done\n")
        if err := w.Close(); err != nil {
                return errors.New("Close: " + err.Error())
        }
        fmt.Printf("Done\n")
        return nil
}

**Expected behavior**
Repeated calls to this routine will work

**Actual behavior**
First call to this routine finishes successfully
Second call hangs in w.Close()

**Screenshots**

**Additional context**

I did a go routine dump on the program when it was hung.
[godump.txt](https://github.com/googleapis/google-cloud-go/files/5626534/godump.txt)

**Comment by tritone** [comment]
Firstly, I would definitely add a timeout to the context that you pass to `NewWriter` to avoid indefinite hangs (though obviously this won't fix whatever underlying problem is happening here).

The way you are using the storage client and writer here looks basically fine otherwise, though unfortunately since this is not a full repro, I can't run it myself to debug. Many users have used the same storage client to create writers across multiple goroutines without issues. One thing I'm wondering about is your buffer that you are copying from-- how is that being managed? The fact that close is hanging [here](https://github.com/googleapis/google-cloud-go/blob/storage/v1.12.0/storage/writer.go#L230) indicates to me that the writer might be waiting on more data to send.

**Comment by mfederw** [comment]
Thanks for taking a look at this.  I'll look closer at the buffer management.  I'll plan to update this when I know more.

**Comment by tritone** [comment]
Sounds good, thanks for following up!
